File: guests/js/scrollto.ts
// The scroll-to scene (tools/scenes/scrollto.steps): the app scrolls a
// list of messages to a row by key (docs/scroll-to-plan.md), opening at
// the newest one before the first layout, jumping to one on a click,
// staying put on a key no row holds, and following its own send.
//     KAYA_SELFTEST=scrollto node guests/js/scrollto.ts

import * as kaya from "kaya-gui";

const Message = kaya.record({ text: String }, "Message");

const app = new kaya.App();

let sent = 60;

function onJump(): void {
  list.scrollToRow("m10");
}

function onNowhere(): void {
  list.scrollToRow("m999");
}

function onSend(): void {
  sent += 1;
  messages.insert(`m${sent}`, Message({ text: `message ${sent}` }));
  count.set(`${sent} messages`);
  list.scrollToRow(`m${sent}`);
}

const { messages, list, count } = app.window(() => {
  const messages = kaya.collection(Message);
  const count = kaya.signal("60 messages");

  const list = kaya.column(() => {
    kaya.label({ bind: count }).a11yId("count");
    kaya.row(() => {
      kaya.button("jump", { onClick: onJump }).a11yId("jump");
      kaya.button("nowhere", { onClick: onNowhere }).a11yId("nowhere");
      kaya.button("send", { onClick: onSend }).a11yId("send");
    });
    return kaya.scroll({ grow: 1 }, () => {
      // The For's own container is what a scrollToRow addresses
      // (docs/scroll-to-plan.md S1): its rows loop names it.
      const rows = messages.rows({ a11yId: "messages" });
      for (const message of rows) {
        kaya.label({ bind: message.text });
      }
      return rows.container;
    });
  });

  for (let i = 1; i <= 60; i++) {
    messages.insert(`m${i}`, Message({ text: `message ${i}` }));
  }
  list.scrollToRow("m60");
  return { messages, list, count };
});

app.run();

File: guests/python/scrollto.py
"""The scroll-to scene, Python port (tools/scenes/scrollto.steps): the app
scrolls a list of messages to a row by key (docs/scroll-to-plan.md),
opening at the newest one before the first layout, jumping to one on a
click, staying put on a key no row holds, and following its own send.
See guests/rust/scrollto.rs."""

import sys
from dataclasses import dataclass

import kaya


@dataclass
class Message:
    text: str


app = kaya.App()
sent = 60


def on_jump():
    messages.container().scroll_to_row("m10")


def on_nowhere():
    messages.container().scroll_to_row("m999")


def on_send():
    global sent
    sent += 1
    messages.insert(f"m{sent}", Message(text=f"message {sent}"))
    count.set(f"{sent} messages")
    messages.container().scroll_to_row(f"m{sent}")


with app.window():
    messages = kaya.collection(Message)
    count = kaya.signal("60 messages")

    with kaya.column():
        kaya.label(bind=count).a11y_id("count")
        with kaya.row():
            kaya.button("jump", on_click=on_jump).a11y_id("jump")
            kaya.button("nowhere", on_click=on_nowhere).a11y_id("nowhere")
            kaya.button("send", on_click=on_send).a11y_id("send")
        with kaya.scroll(grow=1):
            # The For's own container is what a scroll_to_row addresses
            # (docs/scroll-to-plan.md S1): container() names it.
            for message in messages.rows(a11y_id="messages"):
                kaya.label(bind=message.text)

    for i in range(1, 61):
        messages.insert(f"m{i}", Message(text=f"message {i}"))
    messages.container().scroll_to_row("m60")

sys.exit(app.run())
